GRAPHICS/SetParticleFxNonLoopedColour.md

```diff
@@ -8,13 +8,13 @@ ns: GRAPHICS
 void SET_PARTICLE_FX_NON_LOOPED_COLOUR(float r, float g, float b);
 ```
 
-Only works on some fx's, while on others it might SEEM to work "properly", but the colors can be "strange" or even completly different from what you've expected. Reason for this is that those fx's might already have colors "baked into them" which then start to act as a "mixing palette", resulting in a different color than expected. A hypothetical example of this would be if the fx itself is already full (bright) red (RGB: 1.0, 0.0, 0.0) and you then set the color to (bright) green (RGB: 0.0, 1.0, 0.0), that it MIGHT result in Yellow (RGB: 1.0, 1.0, 0.0).
+Only works on some fx's, while on others it might SEEM to work "properly", but the colors can be "strange" or even completly different from what you've expected. Reason for this is that those fx's might already have colors "baked into them" which then start to act as a "mixing palette", resulting in a different color than expected. A hypothetical example of this would be if the fx itself is already full (bright) red (RGB: 255.0, 0.0, 0.0) and you then set the color to (bright) green (RGB: 0.0, 255.0, 0.0), that it MIGHT result in Yellow (RGB: 255.0, 255.0, 0.0).
```

Comment:
I'm sorry and correct me if I'm wrong, but this just seems wrong to me. I disassembled the function in-game and got different results.

I would love to know how you came up with this :-)
Let's give this a rundown over what I did:

1. I looked up the native (0x26143A59EF48B262) on a game dump.
2. I noticed this: ![img](https://i.imgur.com/QBk5K3S.png)
3. I went into the method `sub_7FF703B1FBF8`
4. I noticed the native sets 3 global variables, so I decided to look for them: ![img](https://i.imgur.com/ev6ndED.png)
5. I then noticed other methods set these global variables to `0xBF800000`, which is a IEEE-754 hexadecimal, and when converted to Floating-Point you get `1.0`.
6. I then noticed that this so called globals, also have a getter: ![img](https://i.imgur.com/xXE25Vh.png)
7. Now that getter sets the function parameters so this decompiled function down below can manipulate them, notice how the function multiplies those values by `255.0`. Now if these values you were feeding to the native function were `255.0` as you're describing, the game would multiply those values and the rgb values would be massive.

```c
void __fastcall sub_7FF70368CADC(int *a1, int *a2, __int64 a3, int a4, __int64 a5, __int64 a6, __int64 a7, __int64 a8, float *r, float *g, float *b, float *a9)
{
  __int64 v16; // rbx
  float r_color_copy; // xmm2_4
  int color_r_integer; // ecx
  int v19; // er9
  unsigned __int8 color_r_integer_c; // al
  int c_r_i; // eax
  int color_g_integer; // er8
  char color_g_integer_c; // cl
  unsigned __int8 c_g_i; // dl
  char color_b_integer; // cl
  int color_g_r_combined; // eax
  unsigned __int8 c_b_i; // dl
  unsigned int v28; // eax
  _DWORD *v29; // rax
  __int16 v30; // [rsp+80h] [rbp-28h] BYREF
  char color_red; // [rsp+82h] [rbp-26h] BYREF
  bool v32; // [rsp+83h] [rbp-25h] BYREF
  char v33[4]; // [rsp+84h] [rbp-24h] BYREF
  int v34; // [rsp+88h] [rbp-20h] BYREF
  int v35; // [rsp+8Ch] [rbp-1Ch] BYREF

  v16 = 0i64;
  if ( rage::netEventMgr::CheckForSpaceInPool(g_netEventMgr, 0i64) )
  {
    color_red = 0;
    v30 = 0;
    r_color_copy = *r;
    v33[0] = *r >= 0.0;
    v32 = *a9 >= 0.0;
    if ( r_color_copy >= 0.0 )
    {
      color_r_integer = (int)(float)(r_color_copy * 255.0);
      v19 = (int)(float)(*b * 255.0); // notice here
      color_r_integer_c = color_r_integer;
      if ( color_r_integer > 255 )
        color_r_integer_c = -1;
      if ( color_r_integer < 0 )
        color_r_integer_c = 0;
      c_r_i = color_r_integer_c << 8;
      color_g_integer = (int)(float)(*g * 255.0); // here
      color_g_integer_c = color_g_integer;
      if ( color_g_integer > 255 )
        color_g_integer_c = -1;
      c_g_i = color_g_integer_c;
      if ( color_g_integer < 0 )
        c_g_i = 0;
      color_b_integer = (int)(float)(*b * 255.0); // and here
      color_g_r_combined = (c_g_i | c_r_i) << 8;
      if ( v19 > 255 )
        color_b_integer = -1;
      c_b_i = color_b_integer;
      if ( v19 < 0 )
        c_b_i = 0;
      v28 = c_b_i | color_g_r_combined | 0xFF000000;// completely opaque
      v30 = v28;
      color_red = BYTE2(v28);
    }
    v29 = rage::fwBasePool::New(qword_7FF7050ED028);
    if ( v29 )
    {
      v34 = *a2;
      v35 = *a1;
      v16 = CNetworkPtFXEvent::ctor(
              (__int64)v29,
              &v35,
              &v34,
              a3,
              a4,
              (_DWORD *)a5,
              (_DWORD *)a6,
              (_DWORD *)a7,
              (_BYTE *)a8,
              v33,
              &color_red,                       // red
              (_BYTE *)&v30 + 1,                // green
              (char *)&v30,                     // blue
              (char *)&v32,
              a9);
    }
    CNetworkEventMgr::_addEvent(g_netEventMgr, v16);
  }
}
```